File: Data Structures & Algorithms/validate-parentheses/submission-2.java
class Solution {
    public boolean isValid(String s) {
        int n=s.length();
        Stack<Character> st=new Stack<>();
        Map<Character,Character> mp=new HashMap<>();
        mp.put(')','(');
        mp.put('}','{');
        mp.put(']','[');
        for(char c:s.toCharArray())
        {
            
            if(c=='(' || c=='{' || c=='['){
                st.push(c);
                System.out.println("char : "+c);
            }
            else
            {
                if(!st.isEmpty() && st.peek() == mp.get(c))
                {
                    
                        System.out.println("char : "+c);
                        System.out.println("char : "+st.peek());                    
                        st.pop();
                }
                else
                    return false;  
            }
        }
        return (st.isEmpty())?true:false;
    }
}

File: Data Structures & Algorithms/valid-parenthesis-string/submission-2.py
class Solution:
    def checkValidString(self, s: str) -> bool:
        n = len(s)

        

        st =[]
        star= []

        for i,c in enumerate(s):

            if c == '(':
                st.append(i)
            elif c == '*':
                star.append(i)
            else:
                if st:
                    st.pop()
                elif star:
                    star.pop()
                else:
                    return False
        
        while st and star:
            if st[-1] < star[-1]:
                st.pop()
                star.pop()
            else:
                return False
        
        return len(st) == 0
            

        